pyomo/gdp/tests/test_mbigm.py

```diff
@@ -0,0 +1,684 @@
+                             (m.d1.x1_bounds, m.d3),
+                             (m.d1.x2_bounds, m.d3),
+                             (m.d2.x1_bounds, m.d1),
+                             (m.d2.x2_bounds, m.d1),
+                             (m.d2.x1_bounds, m.d3),
+                             (m.d2.x2_bounds, m.d3),
+                             (m.d3.x1_bounds, m.d1),
+                             (m.d3.x2_bounds, m.d1),
+                             (m.d3.x1_bounds, m.d2),
+                             (m.d3.x2_bounds, m.d2)]:
+            self.assertIn("(%s, %s)" % (cons.name, disj.name), warnings)
+
+        # check that the bounds constraints are right
+        # for x1:
+        cons = mbm.get_transformed_constraints(m.d1.x1_bounds)
+        self.assertEqual(len(cons), 2)
+        sameish = mbm.get_transformed_constraints(m.d4.x1_ub)
+        self.assertEqual(len(sameish), 1)
+        self.assertIs(sameish[0], cons[1])
+
+        self.check_pretty_bound_constraints(cons[1], m.x1, {m.d1: 2, m.d2: 3,
+                                                            m.d3: 10, m.d4: 8},
+                                            lb=False)
+        self.check_pretty_bound_constraints(cons[0], m.x1, {m.d1: 0.5, m.d2:
+                                                            0.65, m.d3: 2, m.d4:
+                                                            -10}, lb=True)
+
+        # and for x2:
+        cons = mbm.get_transformed_constraints(m.d1.x2_bounds)
+        self.assertEqual(len(cons), 2)
+        sameish = mbm.get_transformed_constraints(m.d4.x2_lb)
+        self.assertEqual(len(sameish), 1)
+        self.assertIs(sameish[0], cons[0])
+
+        self.check_pretty_bound_constraints(cons[1], m.x2, {m.d1: 3, m.d2: 10,
+                                                            m.d3: 1, m.d4: 20},
+                                            lb=False)
+        self.check_pretty_bound_constraints(cons[0], m.x2, {m.d1: 0.75, m.d2: 3,
+                                                            m.d3: 0.55, m.d4:
+                                                            -5}, lb=True)
+
+    @unittest.skipUnless(gurobi_available, "Gurobi is not available")
+    def test_nested_gdp(self):
+        # We'll just put another disjunction on one of our disjuncts to test
+        # this. There is little that is special here--we transform from leaf to
+        # root, so the child Disjuncts do not interfere in the calculation of
+        # the M values for their parents. However, their transformed constraints
+        # are placed above the parent, so note that this means that for certain
+        # models, it would be beneficial to flatten the model before using this
+        # transformation, if the children *do* provide insight into the bounds
+        # of the parent. (One could use them and in the subproblems and then
+        # move them up, but that means the M-value subproblems would be
+        # MI(N)LPs, which is a downside.)
```

Comment:
Shouldn't the constraints from the leaf disjuncts be included in their parent disjunction? That would mean they only interfere within that disjunct (contrary to leaving them above the parent). In most cases, moving the leaf disjuncts' reformulated constraints to the root of the model will lead to unnecessary binary variables, but I can imagine an example where it might be invalid for the other parent disjunctions, e.g., [1 <= x <= 3, [1 <= y <= 5] v [6 <= y <= 10]] v [5 <= x <= 10, 15 <= y <= 20]. Moving the inner disjunction constraints to the root and computing bounds based on it for the second parent disjunct leads to infeasibility.

Reply:
Ah, oops, that's a very good point. Because this transformation assumes the exactly-one constraint will be enforced and it isn't necessarily going to be for nested disjunctions. I think you're right that the safest thing to do is just to put transformed constraints on the parent Disjunct. That probably means that flattening the model will often lead to a better relaxation for this transformation...

Reply:
Actually, on second thought, I really don't think this transformation makes sense for nested GDPs. Because if I did what I said above, the calculated M values would depend on the order the Disjuncts are transformed in--that is, whether or not you use the inner disjunctions' information. I can construct examples where this is weaker than flattening the GDP first, so I don't know why you would do it.

Reply:
So you are proposing to flatten the disjunction first?

Reply:
I'm really proposing to punt. I think flattening makes more sense than the leaf-to-root version. So unless we can think of a use case for the latter, I'm thinking I don't implement it.